sdks/python/apache_beam/testing/test_stream_test.py

```diff
@@ -125,6 +125,98 @@ def process(self, element=beam.DoFn.ElementParam,
 
     p.run()
 
+  def test_multiple_outputs(self):
+    """Tests that the TestStream supports emitting to multiple PCollections."""
+    test_stream = (TestStream()
+                   .advance_watermark_to(5, tag='letters')
+                   .add_elements(['a', 'b', 'c'], tag='letters')
+                   .advance_watermark_to(10, tag='numbers')
+                   .add_elements(['1', '2', '3'], tag='numbers'))
+
+    class RecordFn(beam.DoFn):
+      def process(self, element=beam.DoFn.ElementParam,
+                  timestamp=beam.DoFn.TimestampParam):
+        yield (element, timestamp)
+
+    options = StandardOptions(streaming=True)
+    p = TestPipeline(options=options)
+
+    main = p | test_stream
+    letters = main['letters'] | 'record letters' >> beam.ParDo(RecordFn())
+    numbers = main['numbers'] | 'record numbers' >> beam.ParDo(RecordFn())
+
+    assert_that(letters, equal_to([
+        ('a', Timestamp(5)),
+        ('b', Timestamp(5)),
+        ('c', Timestamp(5))]), label='assert letters')
+
+    assert_that(numbers, equal_to([
+        ('1', Timestamp(10)),
+        ('2', Timestamp(10)),
+        ('3', Timestamp(10))]), label='assert numbers')
+
+    p.run()
+
+  def test_multiple_outputs_with_watermark_advancement(self):
+    """Tests that the TestStream can independently output watermarks."""
+
+    # Purposely set the watmerk of numbers to 20 then letters to 5 to test
+    # that the watermark advancement is per PCollection.
+    test_stream = (TestStream()
+                   .advance_watermark_to(20, tag='numbers')
+                   .advance_watermark_to(5, tag='letters')
+                   .add_elements(['a', 'b', 'c'], tag='letters')
+                   .advance_watermark_to(10, tag='letters')
+                   .add_elements(['1', '2', '3'], tag='numbers')
+                   .advance_watermark_to(30, tag='numbers'))
+
+    options = StandardOptions(streaming=True)
+    p = TestPipeline(options=options)
+
+    main = p | test_stream
+    letters = (main['letters']
+               | 'letter windows' >> beam.WindowInto(
+                   FixedWindows(15),
+                   trigger=trigger.AfterWatermark(early=trigger.AfterCount(1)),
+                   accumulation_mode=trigger.AccumulationMode.DISCARDING)
+               | 'letter with key' >> beam.Map(lambda x: ('k', x))
+               | 'letter gbk' >> beam.GroupByKey())
+
+    numbers = (main['numbers']
+               | 'number windows' >> beam.WindowInto(
+                   FixedWindows(15),
+                   trigger=trigger.AfterWatermark(early=trigger.AfterCount(1)),
+                   accumulation_mode=trigger.AccumulationMode.DISCARDING)
+               | 'number with key' >> beam.Map(lambda x: ('k', x))
+               | 'number gbk' >> beam.GroupByKey())
+
+    expected_letters = {
+        window.IntervalWindow(0, 15): [
+            ('k', ['a', 'b', 'c']),
```

Comment:
Thanks for adding this test to verify the watermarks! I assume this line is for the trigger because of the advancement of watermark to 5 for "letters", and then the following line is an empty pane because we are discarding after the trigger because of the accumulation mode. Please add comments to help the reader understand this test.

Reply:
Cool, done.